### ISSUE TITLE
BackupSession fails to be created on Openshift 3.11

### ISSUE THREAD

**Issue by zellerfagyi** [stashed/stash#1080]
Hi,
created BackupConfiguration (and all other objects) according to official guide, CronJob got created successfully from it automatically, but when time has come to run the backup job, it fails to run correctly and the following logs are seen in the backup pods:

```
I0422 08:45:46.175354       1 log.go:172] FLAG: --alsologtostderr="false"
I0422 08:45:46.175718       1 log.go:172] FLAG: --bypass-validating-webhook-xray="false"
I0422 08:45:46.175748       1 log.go:172] FLAG: --enable-analytics="true"
I0422 08:45:46.175757       1 log.go:172] FLAG: --help="false"
I0422 08:45:46.175771       1 log.go:172] FLAG: --invoker-name="my-postgres-backup"
I0422 08:45:46.175780       1 log.go:172] FLAG: --invoker-type="BackupConfiguration"
I0422 08:45:46.175796       1 log.go:172] FLAG: --kubeconfig=""
I0422 08:45:46.175811       1 log.go:172] FLAG: --log-flush-frequency="5s"
I0422 08:45:46.175833       1 log.go:172] FLAG: --log_backtrace_at=":0"
I0422 08:45:46.175845       1 log.go:172] FLAG: --log_dir=""
I0422 08:45:46.175856       1 log.go:172] FLAG: --logtostderr="true"
I0422 08:45:46.175863       1 log.go:172] FLAG: --master=""
I0422 08:45:46.175872       1 log.go:172] FLAG: --service-name="stash-operator"
I0422 08:45:46.175884       1 log.go:172] FLAG: --stderrthreshold="0"
I0422 08:45:46.175895       1 log.go:172] FLAG: --use-kubeapiserver-fqdn-for-aks="true"
I0422 08:45:46.175911       1 log.go:172] FLAG: --v="0"
I0422 08:45:46.175923       1 log.go:172] FLAG: --vmodule=""
W0422 08:45:46.254104       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
F0422 08:45:46.327508       1 create_backupsession.go:79] <nil>
```

Now, apparently there is a ServiceAccount created in the namespace I'm running the backup in, bound to _stash-cron-job_ ClusterRole:
```
$ kubectl get rolebinding -n mynamespace  -l app.kubernetes.io/component=stash-backup
NAME                                  ROLE                USERS     GROUPS    SERVICE ACCOUNTS                      SUBJECTS
my-postgres-backup     /stash-cron-job                        my-postgres-backup     
my-postgres-backup-0   /stash-backup-job                       my-postgres-backup-0 
```

with the following content
```
apiVersion: authorization.openshift.io/v1
kind: ClusterRole
metadata:
  creationTimestamp: 2020-04-21T09:21:15Z
  labels:
    app.kubernetes.io/component: stash-backup
    app.kubernetes.io/managed-by: stash.appscode.com
  name: stash-cron-job
  resourceVersion: "105118333"
  selfLink: /apis/authorization.openshift.io/v1/clusterroles/stash-cron-job
  uid: 73686d74-83b1-11ea-a9fd-4cd98f5479cd
rules:
- apiGroups:
  - stash.appscode.com
  attributeRestrictions: null
  resources:
  - backupsessions
  verbs:
  - '*'
- apiGroups:
  - stash.appscode.com
  attributeRestrictions: null
  resources:
  - backupconfigurations
  verbs:
  - '*'
- apiGroups:
  - stash.appscode.com
  attributeRestrictions: null
  resources:
  - backupbatches
  verbs:
  - '*'
- apiGroups:
  - ""
  attributeRestrictions: null
  resources:
  - events
  verbs:
  - create
- apiGroups:
  - policy
  attributeRestrictions: null
  resourceNames:
  - stash-backupsession-cron
  resources:
  - podsecuritypolicies
  verbs:
  - use
- apiGroups:
  - apps
  attributeRestrictions: null
  resources:
  - daemonsets
  - deployments
  - replicasets
  - statefulsets
  verbs:
  - get
- apiGroups:
  - ""
  attributeRestrictions: null
  resources:
  - persistentvolumeclaims
  - replicationcontrollers
  verbs:
  - get
- apiGroups:
  - apps.openshift.io
  attributeRestrictions: null
  resources:
  - deploymentconfigs
  verbs:
  - get
- apiGroups:
  - appcatalog.appscode.com
  attributeRestrictions: null
  resources:
  - '*'
  verbs:
  - get
```

If I replace 
```
- apiGroups:
  - stash.appscode.com
  attributeRestrictions: null
  resources:
  - backupsessions
  verbs:
  - '*'
- apiGroups:
  - stash.appscode.com
  attributeRestrictions: null
  resources:
  - backupconfigurations
  verbs:
  - '*'
- apiGroups:
  - stash.appscode.com
  attributeRestrictions: null
  resources:
  - backupbatches
  verbs:
  - '*'
```

with
```
- apiGroups:
  - stash.appscode.com
  attributeRestrictions: null
  resources:
  - '*'
  verbs:
  - '*'
```

then the BackupSession can be created correctly automatically by the pod created from CronJob.
The problem is, that every time I modify the corresponding BackupConfiguration (e.g. to re-schedule it), it automatically patches the stash-cron-job ClusterRole and effectively resets to the original (non-working) permissions, snippet from _stash-operator_ pod's _operator_ container:

```
I0422 08:52:36.544146       1 clusterrole.go:72] Patching ClusterRole stash-cron-job with {"rules":[{"apiGroups":["stash.appscode.com"],"resources":["backupsessions"],"verbs":["*"]},{"apiGroups":["stash.appscode.com"],"resources":["backupconfigurations"],"verbs":["*"]},{"apiGroups":["stash.appscode.com"],"resources":["backupbatches"],"verbs":["*"]},{"apiGroups":[""],"resources":["events"],"verbs":["create"]},{"apiGroups":["policy"],"resourceNames":["stash-backupsession-cron"],"resources":["podsecuritypolicies"],"verbs":["use"]},{"apiGroups":["apps"],"resources":["deployments","statefulsets","replicasets","daemonsets"],"verbs":["get"]},{"apiGroups":[""],"resources":["replicationcontrollers","persistentvolumeclaims"],"verbs":["get"]},{"apiGroups":["apps.openshift.io"],"resources":["deploymentconfigs"],"verbs":["get"]},{"apiGroups":["appcatalog.appscode.com"],"resources":["*"],"verbs":["get"]}]}.
```

As a workaround, I used following (crude) method to allow BackupSession to be created:
- modify BackupConfiguration's schedule to fit my needs
- then edit _stash-cron-job_  ClusterRole to allow '*' resources in apiGroup: stash.appscode.com instead of the granular backupsessions, backupconfigurations, backupbatches, which seem not to be enough

Could you please help with what the problem might be on Openshift3; why BackupSession cannot be created automatically with the default stash.appscode.com apiGroups permissions? 
Thank you!

**Comment by hossainemruz** [comment]
What version of Stash are you using?

**Comment by zellerfagyi** [comment]
sorry for not including it in my initial submission, its `v0.9.0-rc.6` 
```
I0421 11:13:02.777018       1 run.go:42] Starting operator version v0.9.0-rc.6+473dab81bfc0d3f8085b2fbf99c03143501d991d ...
```

**Comment by hossainemruz** [comment]
Can you try the latest image.

```
helm install stash-operator appscode/stash \
  --version v0.9.0-rc.6 \
  --namespace kube-system \
  --set operator.registry=appscodeci \
  --set operator.tag=v0.9.0-rc.6-5-g5cdeebee_linux_amd64
```
The image was built after fixing #1069 . This will give us the proper error message.

I believe this is happening because the CronJob is failing to set OwnerReference to the BackupSession. Probably missing permission  for `backupsessions/finalizers`

**Comment by zellerfagyi** [comment]
thanks for the hint! redeployed Stash with the image suggested
```
I0422 13:08:15.634107       1 run.go:42] Starting operator version v0.9.0-rc.6-5-g5cdeebee+5cdeebeec314c140606f4a45dace5d0903bff5b8 ...
```

and indeed it now has a more insightful log on the permission issue, seems like the one you hinted (create-backupsession pod log):
```
I0422 13:15:14.392851       1 log.go:172] FLAG: --alsologtostderr="false"
I0422 13:15:14.393031       1 log.go:172] FLAG: --bypass-validating-webhook-xray="false"
I0422 13:15:14.393054       1 log.go:172] FLAG: --enable-analytics="true"
I0422 13:15:14.393058       1 log.go:172] FLAG: --help="false"
I0422 13:15:14.393066       1 log.go:172] FLAG: --invoker-name="my-postgres"
I0422 13:15:14.393070       1 log.go:172] FLAG: --invoker-type="BackupConfiguration"
I0422 13:15:14.393075       1 log.go:172] FLAG: --kubeconfig=""
I0422 13:15:14.393079       1 log.go:172] FLAG: --log-flush-frequency="5s"
I0422 13:15:14.393092       1 log.go:172] FLAG: --log_backtrace_at=":0"
I0422 13:15:14.393097       1 log.go:172] FLAG: --log_dir=""
I0422 13:15:14.393102       1 log.go:172] FLAG: --logtostderr="true"
I0422 13:15:14.393106       1 log.go:172] FLAG: --master=""
I0422 13:15:14.393110       1 log.go:172] FLAG: --service-name="stash-operator"
I0422 13:15:14.393118       1 log.go:172] FLAG: --stderrthreshold="0"
I0422 13:15:14.393124       1 log.go:172] FLAG: --use-kubeapiserver-fqdn-for-aks="true"
I0422 13:15:14.393147       1 log.go:172] FLAG: --v="0"
I0422 13:15:14.393153       1 log.go:172] FLAG: --vmodule=""
W0422 13:15:14.471134       1 client_config.go:541] Neither --kubeconfig nor --master was specified.  Using the inClusterConfig.  This might not work.
F0422 13:15:14.552417       1 create_backupsession.go:79] backupsessions.stash.appscode.com "my-postgres-1587561314" is forbidden: cannot set blockOwnerDeletion if an ownerReference refers to a resource you can't set finalizers on: no RBAC policy matched, <nil>
```

**Comment by hossainemruz** [comment]
So, you just need to add `update` permission of `backupsessions/finalizers` in the CronJob's clusterrole. We are going to fix it soon.

**Comment by zellerfagyi** [comment]
thank you for identifying the problem! sadly the patched ClusterRole is always overridden to the default which misses `backupsessions/finalizers update` permission when BackupConfiguration is modified (see in my original submission.)

Guess we'll need to wait for the fix and again thank you for your quick help !

**Comment by hossainemruz** [comment]
I am going to release a patched image tomorrow after some testing.

**Comment by hossainemruz** [comment]
This issue should be fix by #1083. Can you please confirm?
```
helm install stash-operator appscode/stash \
  --version v0.9.0-rc.6 \
  --namespace kube-system \
  --set operator.registry=appscodeci \
  --set operator.tag=fix-cronjob-rbac_linux_amd64
```

**Comment by zellerfagyi** [comment]
redeployed Stash with the nex image

```
I0423 09:28:07.366397       1 run.go:42] Starting operator version fix-cronjob-rbac+5cdeebeec314c140606f4a45dace5d0903bff5b8 ...
```

then created a BackupConfiguration as usual, and I can confirm that thanks to the fix, CronJob ClusterRole now gets all the permissions for `stash.appscode.com` api resources, not just the granular backupsessions, backupconfigurations, backupbatches:
```
I0423 09:31:22.234070       1 clusterrole.go:72] Patching ClusterRole stash-cron-job with {"rules":[{"apiGroups":["stash.appscode.com"],"resources":["*"],"verbs":["*"]},{"apiGroups":[""],"resources":["events"],"verbs":["create"]},{"apiGroups":["policy"],"resourceNames":["stash-backupsession-cron"],"resources":["podsecuritypolicies"],"verbs":["use"]},{"apiGroups":["apps"],"resources":["deployments","statefulsets","replicasets","daemonsets"],"verbs":["get"]},{"apiGroups":[""],"resources":["replicationcontrollers","persistentvolumeclaims"],"verbs":["get"]},{"apiGroups":["apps.openshift.io"],"resources":["deploymentconfigs"],"verbs":["get"]},{"apiGroups":["appcatalog.appscode.com"],"resources":["*"],"verbs":["get"]}]}.
```

which allows BackupSession now to be correctly created
```
$ kubectl get backupsession -n mynamespace 
NAME                       AGE
my-postgres-1587634513   20s
```

Fix is confirmed from my part, thank you very much for fixing it so quickly!
